### PR TITLE
Add viewport detection using the Verge module

### LINKED ISSUE #12: Ad in viewport detection
An ad should only trigger a "view" when the ad is in the viewport. For most publishers, this isn't a problem since the ad is above the fold and visible and this absolutely should be a requirement for publishers. However, we should also verify this. Publishers may neglect to test their ad placement on different screen sizes or on mobile.

I was pretty certain that the IAB sets out these rules in their [terms](https://www.iab.com/guidelines/standard-terms-conditions-internet-advertising-media-buys-one-year-less/) but I couldn't find the relevant section or possibly it's in another document.

RTD does [viewport detection](https://github.com/readthedocs/readthedocs.org/blob/505be8cc8fabc0dbdf290336dd263c887ede459b/readthedocs/core/static-src/core/js/doc-embed/sponsorship.js#L181-L195) and uses the [verge](https://github.com/ryanve/verge) module to do so.

### PR DESCRIPTION
**Pull request by davidfischer**
- Uses the same detection method as RTD uses
- Adds a negligible ~1.5kb minified but not gzipped

One aspect of this that I don't like is that it uses setInterval to check whether the ad is in the viewport. This could be a minor performance hit although in most cases it shouldn't be since it stops checking once it sees the ad in the viewport.

In terms of rollout, it might be worth testing this on some of our biggest publishers just to sanity check.

Fixes #12.

### CODE REVIEW

**Comment by davidfischer** [comment]
I should also mention that currently the view pixel is included in the HTML response directly. That would need to be removed but that can be changed server side after this rolls out.

File: index.js [hunk 1]
```diff
@@ -99,10 +102,27 @@ export class Placement {
...
+        if (placement.response && placement.response.view_url && verge.inViewport(element, -3)) {
+          // This ad was seen!
+          var pixel = document.createElement("img");
+          pixel.src = placement.response.view_url;
+          pixel.className = "ea-pixel";
+          element.appendChild(pixel);
+
+          clearInterval(viewport_detection);
+        }
+      }, 100, placement.target);
```
Comment (ericholscher) [inline index.js:124]: I assume this is 100ms? Seems reasonable, but we could go even lower if we wanted probably if we're worried about it.

**Review by ericholscher — APPROVED** [review]
This seems pretty straightforward. We should definitely do some testing on it -- I wonder if there's a way for us to feature flag it? I guess the other way if to just keep the current implementation, and test this live while generating duplicate pixel views, but we already account for that on the server.

File: index.js (hunk 1 above)
Comment (davidfischer) [reply index.js:124]: It is 100ms. I don't know if we really need to go lower. I believe the IAB guidelines are that an ad should be visible for a full second although I can't seem to find that guideline right now. One possibility is to both trigger it immediately and every 100ms.

**Comment by davidfischer** [comment]
> I guess the other way if to just keep the current implementation, and test this live while generating duplicate pixel views, but we already account for that on the server.

This essentially was my plan.

File: index.js (hunk 1 above)
Comment (davidfischer) [reply index.js:124]: I did end up finding it although there's conflicting guidelines.

Firstly there's this ([source](https://www.iab.com/wp-content/uploads/2015/06/MRC-Viewable-Ad-Impression-Measurement-Guideline.pdf)):

> Requirements for Viewable Display Advertising Impressions
In addition to the above requirements, Viewable Display Ad Impressions are
counted when the following criteria are met:
> • Pixel Requirement: Greater than or equal to 50% of the pixels in the
advertisement were on an in-focus browser tab on the viewable space of
the browser page, and
> • Time Requirement: The time the pixel requirement is met was greater
than or equal to one continuous second, post ad render.

Secondly there's this ([source](https://www.iab.com/wp-content/uploads/2017/11/Desktop-Display-Impression-Measurement-Guidelines-US-MMTF-Final-v1.1.pdf)):

> An Ad Impression across all display marketing channels is the measurement of responses
from an ad delivery system to an ad request from the user's browser, which is filtered for
invalid traffic and is recorded at a point as late as possible in the process of delivery of the
creative material to the user's browser. The ad must be loaded and at minimum begin to
render in order to count it as a valid ad impression. 

The latter of these is more recent and seems more correctly applicable.

Comment (ericholscher) [reply index.js:124]: Yea, it's better than what we have now either way :)

File: index.js [hunk 2]
```diff
@@ -99,10 +102,27 @@ export class Placement {
       this.target.appendChild(element);
 
       return this;
+    }).then((placement) => {
+      // Detect when the ad is in the viewport
+      // Add the view pixel to the DOM to count the view
+
+      var viewport_detection = setInterval((element) => {
```
Comment (agjohnson) [inline index.js:109]: While `var` works here, `let` reduces the scope of the variable, which would be function level with `var` -- `let` is block level scoped.

```suggestion
      let viewport_detection = setInterval((element) => {
```

File: index.js [hunk 3]
```diff
@@ -99,10 +102,27 @@ export class Placement {
...
+        // Verge can be off by 1-2 pixels
+        // A fudge factor of ~3 is needed for the case where the ad
+        // is hidden off the side of the screen by a sliding sidebar
+        // For example, if the right side of the ad is at x=0
+        // or the left side of the ad is at the right side of the viewport
+        if (placement.response && placement.response.view_url && verge.inViewport(element, -3)) {
+          // This ad was seen!
+          var pixel = document.createElement("img");
```
Comment (agjohnson) [inline index.js:117]: Same with block level scope here

**Review by agjohnson — COMMENTED** [review]
Just noting some minor es6 scope optimizations

**Comment by davidfischer** [comment]
I have tested this and it tests out fairly well. I have also deployed it [here](https://media.ethicalads.io/media/client/beta/ethicalads.min.js) in case any other publishers want to try it out. We are using it on our own landing pages.

After reviewing a few dozen integrations of our existing publishers, many of them are doing something custom (like hitting the API directly instead of using the client) which is going to complicate the rollout of this. My plan is to roll out this beta on Read the Docs and if everything goes smoothly there we can proceed with notifying publishers about changes.

**Review by ericholscher — CHANGES_REQUESTED** [review]
Unapproving this, since we aren't able to ship it widely until we do a bunch more work on it. Will be useful once we have it tho 👍

**Review by ericholscher — APPROVED** [review]
I think we are ready to ship this along with https://github.com/readthedocs/ethical-ad-server/pull/269 -- the plan for migrating users to this code is there.

**Pull request merged**